### PR TITLE
support output-dir when running 'helm template'

### PR DESCRIPTION
**Pull request by rajatjindal**
if --output-dir provided, helm template will create the files instead of printing on stdout.

while doing so, the directory structure of the output files will be same as directory structure of input template files.

### CODE REVIEW

**Comment by rajatjindal** [comment]
@technosophos can you please review.

**Comment by rajatjindal** [comment]
cc'ing @bacongobbler also - didn't realize he pinged on helm-users suggesting to open PR with helm project

**Comment by bacongobbler** [comment]
Thanks! We'll take a look at this eventually. We've got a lot on our plates at this time, but we'll come back around when we're doing another round of reviews.

**Comment by rajatjindal** [comment]
sure, thank you

File: cmd/helm/template.go [hunk 1]
```diff
@@ -248,8 +263,53 @@ func (t *templateCmd) run(cmd *cobra.Command, args []string) error {
...
+			continue
+		}
 		fmt.Printf("---\n# Source: %s\n", m.Name)
 		fmt.Println(data)
 	}
 	return nil
 }
+
+// write the <data> to <output-dir>/<name>
+func writeToFile(outputDir string, name string, data string) {
```
Comment (technosophos) [inline cmd/helm/template.go]: Can we have this return an `error` and not `panic` when the create/write fails?

File: cmd/helm/template.go [hunk 2]
```diff
@@ -126,6 +133,14 @@ func (t *templateCmd) run(cmd *cobra.Command, args []string) error {
 		}
 	}
 
+	// verify that output-dir exists if provided
+	if t.outputDir != "" {
+		_, err = os.Stat(t.outputDir)
+		if os.IsNotExist(err) {
+			panic(fmt.Sprintf("output-dir '%s' does not exist", t.outputDir))
```
Comment (technosophos) [inline cmd/helm/template.go]: This should return an error like the `fmt.Errorf()` call a few lines above.

File: cmd/helm/template.go [hunk 3]
```diff
@@ -248,8 +263,53 @@ func (t *templateCmd) run(cmd *cobra.Command, args []string) error {
...
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("wrote %s\n", outfileName)
+	return
+}
+
+// check if the directory exists to create file. creates if don't exists
+func ensureDirectoryForFile(file string) {
```
Comment (technosophos) [inline cmd/helm/template.go]: Again, just returning an error instead of panicing is our preference.

**Review by technosophos — CHANGES_REQUESTED** [review]
This is looking great. I left a few comments to remove `panic`. For the most part, our coding convention is to avoid panic unless something really, really bad happens. So just returning `error` objects is preferable in most cases.

Otherwise, this looks great.

**Review by technosophos — APPROVED** [review]
Updated PR LGTM! Thanks for making the changes.

**Pull request merged**